api/logs/s3/s3.go

```diff
@@ -173,6 +175,245 @@ func (s *store) GetLog(ctx context.Context, appID, callID string) (io.Reader, er
+	ctx, span := trace.StartSpan(ctx, "s3_get_calls")
+	defer span.End()
+
+	if filter.AppID == "" {
+		return nil, errors.New("s3 store does not support listing across all apps")
+	}
+
+	// NOTE:
+	// if filter.Path != ""
+	//   find marker from marker keys, start there, list keys, get next marker from there
+	// else
+	//   use marker for keys
+
+	// NOTE we need marker keys to support (app is REQUIRED):
+	// 1) quick iteration per path
+	// 2) sorted by id across all path
+	// marker key: m : {app} : {path} : {id}
+	// key: s: {app} : {id}
+	//
+	// also s3 api returns sorted in lexicographic order, we need the reverse of this.
+
+	// marker is either a provided marker, or a key we create based on parameters
+	// that contains app_id, may be a marker key if path is provided, and may
+	// have a time guesstimate if to time is provided.
+
+	var marker string
+
+	// filter.Cursor is a call id, translate to our key format. if a path is
+	// provided, we list keys from markers instead.
+	if filter.Cursor != "" {
+		marker = callKey(filter.AppID, filter.Cursor)
+		if filter.Path != "" {
+			marker = callMarkerKey(filter.AppID, filter.Path, filter.Cursor)
+		}
+	} else if t := time.Time(filter.ToTime); !t.IsZero() {
+		// we need to use first 48 bits of id to approximate created_at, to skip over records
+		// if looking far into the past to save time.
+		ms := uint64(t.Unix()*1000) + uint64(t.Nanosecond()/int(time.Millisecond))
+		var buf id.Id
+		buf[0] = byte(ms >> 40)
+		buf[1] = byte(ms >> 32)
+		buf[2] = byte(ms >> 24)
+		buf[3] = byte(ms >> 16)
+		buf[4] = byte(ms >> 8)
+		buf[5] = byte(ms)
+		mid := buf.String()[:10]
+		// timestamp is first 10 bytes of string encoded
+		marker = callKey(filter.AppID, mid)
+		if filter.Path != "" {
+			marker = callMarkerKey(filter.AppID, filter.Path, mid)
+		}
+	}
+
+	// prefix prevents leaving bounds of app or path marker keys
+	prefix := callKey(filter.AppID, "")
+	if filter.Path != "" {
+		prefix = callMarkerKey(filter.AppID, filter.Path, "")
+	}
+
+	input := &s3.ListObjectsInput{
+		Bucket:  aws.String(s.bucket),
+		MaxKeys: aws.Int64(int64(filter.PerPage)),
+		Marker:  aws.String(marker),
+		Prefix:  aws.String(prefix),
+	}
+
+	result, err := s.client.ListObjects(input)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list logs: %v", err)
+	}
+
+	calls := make([]*models.Call, 0, len(result.Contents))
+
+	for _, obj := range result.Contents {
+		if len(calls) == filter.PerPage {
+			break
+		}
+
+		// extract the app and id from the key to lookup the object, this also
+		// validates we aren't reading strangely keyed objects from the bucket.
+		var app, id string
+		if filter.Path != "" {
+			fields := strings.Split(*obj.Key, "/")
+			if len(fields) != 4 {
+				return calls, fmt.Errorf("invalid key in call markers: %v", *obj.Key)
+			}
+			app = fields[1]
+			id = fields[3]
+		} else {
+			fields := strings.Split(*obj.Key, "/")
+			app = fields[1]
```

Comment:
app/id setting needs to happen after if check?

Reply:
d'oh. thanks